tensorflow/stream_executor/gpu/gpu_executor.h

```diff
@@ -71,11 +73,13 @@ class GpuExecutor : public internal::StreamExecutorInterface {
               const BlockDim &block_dims, const KernelBase &k,
               const KernelArgsArrayBase &args) override;
 
+  // (supported on CUDA only)
```

Comment:
I thought about this, but this change is non-trivial and might not be worth it.

Code outside of  stream_executor does not call routines that are CUDA/ROCm specific, but those routines are called by other routines which are called from outside. Take for example `GpuExecutor::LoadModule`. It gets called from the XLA code, and in turn will call `LoadPtx / LoadCubin / LoadHsaco`. 

If I move the `LoadHsaco` routine to a `ROCmExecutor` class, then I need to introduce a `reinterpret_cast<ROCmExecutor*>(this)->LoadHsaco` in the implementation of `GpuExecutor::LoadModule`. Potentially lot of similar casts will need to be added. 

`LoadHsaco` needs access to the `in_memory_modules_mu_` mutex, which is also used by other routines, so basically we will end up moving all the private members of `GpuExecutor` to `ROCmExecutor / CUDAExecutor` (or declaring some  as `protected` and leaving them in `GpuExecutor`....that will make code difficult to follow and maintain)

 At that point we might as well revert back to having separate classes / header files for `rocm_gpu_executor.h` / `cuda_gpu_executor.h`. But doing so will require undoing some of the header files that we have merged (since they refer to GpuExecutor)

What do you think?

Reply:
I think it's ok for now, but maybe something we want to revisit in the future when the divergence gets larger.